agent/consul/acl_endpoint_test.go

```diff
@@ -637,12 +637,15 @@ func TestACLEndpoint_TokenRead(t *testing.T) {
 
 	testrpc.WaitForLeader(t, s1.RPC, "dc1")
 
-	token, err := upsertTestToken(codec, "root", "dc1")
+	clock := newStoppedClock(time.Now())
```

Comment:
I've written tests with stubbed clocks like this in the past and can see the utility, although they often end up brittle for other reasons IMO (or encoding assumptions about ordering that may not be true in real life and so missing things).

I know Mitchell has expressed a pretty strong aversion to this pattern in general and it's the reason we don't do stuff like this already in Consul and other tools. We should try to work out why this case is different from others (if it is) and/or consider how to ensure this pattern isn't proliferated absolutely everywhere that deals with time in Consul (which is most subsystems).

https://speakerdeck.com/mitchellh/advanced-testing-with-go?slide=63

Reply:
I've always viewed putting sleeps into tests to be a bug to fix, because they tend to be super flaky when the CI machine is underpowered/overloaded, or slow just because they end up doing a lot of nothing. In the latter case it quickly gets annoying if you want to run your full test suite locally after every small change for a sanity check.

Reply:
I agree that sleeps in tests and timing-sensitivity are bad. There is one school of thought that says they are the least bad option available.

The problem is that any clock abstraction like this is leaky - it doesn't actually behave like real time - concurrent goroutines in the program may still have performed multiple operations in different orders between and yet observed both things happen at the same "time". Similarly, anything you happen to forget to stub out that still does a real Sleep/timer, time.Now() can end up behaving differently in test than it ever could in real life.

My experience is that in really simple and trivial cases with very small units and small amount of scope for replacing the clock it can work well. But with moderate complexity, the tests become brittle for different reasons - for example timeouts and delays that use time.After or tickers (even if you stub them to work with respect to your fake time advancing) actually don't due to the machinery of how you implement that being imperfect. Some related anecdotes I found on Google: https://www.reddit.com/r/golang/comments/640vz3/what_approach_do_you_use_to_mock_timenow/

I don't want to make too big a thing of it - this case does seem somewhat reasonable but as perter bourgon mentioned in link above I find the simplest thing to do now is to arrange the code so that any complex time decisions are made in functions that take `now time.Time` (or a ticker chan or duration or whatever) as an argument that you can specifically control within fake clock in tests, and then just compose those and make the wider tests use real time.

Just my thoughts - I'm not set against this as it seems close to the "contained and simple case" but we've avoided the pattern intentionally until now (admittedly with a lot of sadness around timing in tests, but I don't think that could be avoided this way). It might be interesting to see if other gophers in the org agree?

Reply:
`clock` is gone now.